File: android/app/src/main/java/com/desmoines/aipulse/data/remote/ForYouService.kt
package com.desmoines.aipulse.data.remote

import android.util.Log
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Count
import io.github.jan.supabase.postgrest.rpc
// Count helpers from postgrest module
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant
import java.time.temporal.ChronoUnit
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "ForYouService"

/**
 * Client for the For You / Trending Now rails. Mirrors iOS ForYouService.swift.
 * Falls back to "trending" for users with fewer than 5 swipes in last 90 days.
 */
@Singleton
class ForYouService @Inject constructor(
    private val supabaseClient: SupabaseClient?,
) {

    enum class Source { FOR_YOU, TRENDING }

    @Serializable
    data class Recommendation(
        val id: String,
        val title: String? = null,
        val date: String? = null,
        val category: String? = null,
        val venue: String? = null,
        @SerialName("image_url") val imageUrl: String? = null,
        @SerialName("is_featured") val isFeatured: Boolean? = null,
        @SerialName("recommendation_score") val recommendationScore: Double? = null,
        @SerialName("recommendation_reason") val recommendationReason: String? = null,
    )

    @Serializable
    private data class PersonalParams(
        val p_user_lat: Double? = null,
        val p_user_lon: Double? = null,
        val p_limit: Int,
    )

    @Serializable
    private data class TrendingParams(val p_limit: Int)

    data class Result(val source: Source, val items: List<Recommendation>)

    suspend fun fetch(limit: Int = 12): Result {
        val client = supabaseClient ?: return Result(Source.TRENDING, emptyList())

        val userId = try {
            client.auth.currentSessionOrNull()?.user?.id
        } catch (_: Exception) {
            null
        }

        // Probe swipe count when authenticated
        val swipeCount = if (userId != null) {
            try {
                val cutoff = Instant.now().minus(90, ChronoUnit.DAYS).toString()
                val response = client.from("swipe_interactions").select {
                    count(Count.EXACT)
                    filter {
                        eq("user_id", userId)
                        gte("created_at", cutoff)
                    }
                    limit(1)
                }
                response.countOrNull()?.toInt() ?: 0
            } catch (e: Exception) {
                Log.w(TAG, "swipe count probe failed: ${e.message}")
                0
            }
        } else 0

        return try {
            if (swipeCount >= 5) {
                val rows = client.postgrest.rpc(
                    function = "get_personalized_recommendations",
                    parameters = PersonalParams(p_limit = limit),
                ).decodeList<Recommendation>()
                Result(Source.FOR_YOU, rows)
            } else {
                val rows = client.postgrest.rpc(
                    function = "get_trending_events",
                    parameters = TrendingParams(p_limit = limit),
                ).decodeList<Recommendation>()
                Result(Source.TRENDING, rows)
            }
        } catch (e: Exception) {
            Log.w(TAG, "For You / Trending fetch failed: ${e.message}")
            Result(Source.TRENDING, emptyList())
        }
    }
}

File: ios/DesMoinesInsider/Services/ForYouService.swift
import Foundation

/// Client for the For You / Trending Now event rails. Decides which RPC to
/// call based on the user's swipe history (cold-start fallback under 5
/// swipes). IOS-DISCOVER-2026-002.
@MainActor
@Observable
final class ForYouService {
    static let shared = ForYouService()

    enum Source { case forYou, trending }

    struct Recommendation: Identifiable, Decodable, Hashable {
        let id: UUID
        let title: String?
        let date: String?
        let category: String?
        let imageUrl: String?
        let venue: String?
        let isFeatured: Bool?
        let recommendationScore: Double?
        let recommendationReason: String?

        enum CodingKeys: String, CodingKey {
            case id, title, date, category, venue
            case imageUrl = "image_url"
            case isFeatured = "is_featured"
            case recommendationScore = "recommendation_score"
            case recommendationReason = "recommendation_reason"
        }
    }

    private(set) var recommendations: [Recommendation] = []
    private(set) var source: Source = .trending
    private(set) var isLoading = false

    private let supabase = SupabaseService.shared.client

    /// Merges overlapping refreshes into one. ForYouRail calls refresh from
    /// both a `.task` and a manual button, and the `.task` re-fires whenever
    /// the view identity changes, so two calls landing together is ordinary
    /// rather than exceptional.
    private let refreshFlight = SingleFlight()

    private init() {}

    /// Decide which rail to show:
    /// - 5+ recent swipes → For You (personalized)
    /// - else → Trending Now (cold-start fallback)
    func refresh(limit: Int = 12) async {
        await refreshFlight.run { [weak self] in
            await self?.performRefresh(limit: limit)
        }
    }

    /// The refresh itself. Never call this directly - `refresh` is what
    /// guarantees one execution per overlapping group.
    ///
    /// The `guard !isLoading` this replaced dropped the second caller rather
    /// than joining it: that call returned immediately having awaited
    /// nothing, so `await refresh()` did not mean the data had arrived.
    private func performRefresh(limit: Int) async {
        guard let client = supabase else { return }
        isLoading = true
        defer { isLoading = false }

        do {
            // Probe swipe count (only the count, no detail rows)
            let session = try? await client.auth.session
            let userId = session?.user.id

            var swipeCount = 0
            if let userId {
                struct CountWrapper: Decodable { let count: Int }
                let response = try await client
                    .from("swipe_interactions")
                    .select("*", head: false, count: .exact)
                    .eq("user_id", value: userId.uuidString)
                    .gte("created_at", value: ISO8601DateFormatter().string(from: Date().addingTimeInterval(-90 * 86400)))
                    .limit(1)
                    .execute()
                swipeCount = response.count ?? 0
            }

            if swipeCount >= 5 {
                source = .forYou
                struct Params: Encodable {
                    let p_user_lat: Double?
                    let p_user_lon: Double?
                    let p_limit: Int
                }
                let rows: [Recommendation] = try await client
                    .rpc("get_personalized_recommendations", params: Params(
                        p_user_lat: nil,
                        p_user_lon: nil,
                        p_limit: limit,
                    ))
                    .execute()
                    .value
                recommendations = rows
            } else {
                source = .trending
                struct TrendingParams: Encodable { let p_limit: Int }
                let rows: [Recommendation] = try await client
                    .rpc("get_trending_events", params: TrendingParams(p_limit: limit))
                    .execute()
                    .value
                recommendations = rows
            }
        } catch {
            #if DEBUG
            AppLogger.network.warning("ForYouService refresh failed: \(error.localizedDescription)")
            #endif
        }
    }
}
